exampleCourse/serverFilesCourse/r_autograder/pltest.R

```diff
@@ -0,0 +1,157 @@
+## Script Requirements ----
+
+# Note: Packages listed here are found on the PrairieLearn R docker image
+
+# Use testthat to test the student code
+library("testthat")
+
+# Use stringr to simplify string manipulation
+library("stringr")
+
+# Use jsonlite to export the test results under a PrairieLearn format
+library("jsonlite")
+
+## Script hierarchy
+# /grade
+# |- data/
+# |- results/
+# |- run/                # <--- Working directory
+# |- serverFilesCourse/
+# |- shared/
+# |- student/
+# |- tests/
+
+## Run Unit Tests ----
+
+# Automatically test the directory
+# This helps avoid needing to build a package on a per question basis.
+# To understand testing outside of a package, it's mainly documented in an
+# issue ticket: https://github.com/r-lib/testthat/issues/659
+
+# Begin the process of grading each student's code
+# Add a capture statement to avoid allowing students to output
+# content to console.
+log <- capture.output({
+    test_data_as_list <- testthat::test_dir("../tests",
+                                           reporter = testthat::ListReporter)
+})
+
+# Convert from list to a data.frame
+test_data = as.data.frame(test_data_as_list)
+
+## Compute Unit Test Correctness ----
+
+# Retrieve question totals from the test description. To facilitate this,
+# we assume the format is given as...
+#
+# NumericPoints: test description
+#
+# To ensure no issue, we remove all none-numeric values on the left and right
+# of the colon separator.
+
+# Obtain the maximum points
+test_data$max_points = stringr::str_replace(
+    test_data$test,
+    pattern = ".*([:digit:]+).*:.*",
+    replacement = "\\1"
+)
+
+# Convert retrieved maximum points to a numeric value
+test_data$max_points = as.numeric(test_data$max_points)
+
+# Calculate the number of points the student received
+test_data$earned_points = ifelse(test_data$passed >= 1,
+                                 test_data$max_points,
+                                 0.0)
+
+# Now, let's aim to extract the description values
+test_data$description = stringr::str_replace(
+    test_data$test,
+    pattern = ".*[:digit:]+.*:[:space:](.*)",
+    replacement = "\\1"
+)
+
+## Format the Results to Match External Grading Format ----
+
+# Format the results of each inline test.
+pl_test_list_format = function(i, x) {
+
+    pl_message_statement = function(error, fail) {
+        if(isTRUE(error)) {
+            "Your code encounted an error. Please verify it works."
+        } else if(isTRUE(fail)) {
+            "The code ran but did not produce the correct result."
+        } else {
+            "No errors!"
+        }
+    }
+
+    pl_output_statement = function(passed) {
+        condition = if(passed == 1) {
+            "matched"
+        } else {
+            "did not match"
+        }
+        sprintf("Running test...\nYour output %s the expected output.", condition)
```

Comment:
You don’t want to tell them the expected and actual output? Seems like debugging would be easier if they were told.

Reply:
At the moment this information seems very redundant with the message above.

Reply:
Agreed. I'll probably add a new layer to parse the error messages.

Reply:
Looks great now!